File: app/src/main/java/org/neidhardt/dynamicsoundboard/misc/longtermtask/LongTermTask.java
package org.neidhardt.dynamicsoundboard.misc.longtermtask;

import de.greenrobot.event.EventBus;
import org.neidhardt.dynamicsoundboard.misc.Logger;
import org.neidhardt.dynamicsoundboard.misc.longtermtask.events.LongTermTaskStateChangedEvent;
import roboguice.util.SafeAsyncTask;

/**
 * Created by eric.neidhardt on 24.03.2015.
 */

public abstract class LongTermTask<T> extends SafeAsyncTask<T>
{
	private static int taskCounter;

	@Override
	protected void onPreExecute() throws Exception
	{
		super.onPreExecute();
		taskCounter++;
		EventBus.getDefault().postSticky(new LongTermTaskStateChangedEvent(true, taskCounter));
	}

	@Override
	protected void onSuccess(T t) throws Exception
	{
		super.onSuccess(t);
		taskCounter--;
		EventBus.getDefault().postSticky(new LongTermTaskStateChangedEvent(false, taskCounter));
	}

	@Override
	protected void onException(Exception e) throws RuntimeException
	{
		super.onException(e);
		Logger.e(getTag(), e.getMessage());
		taskCounter--;
		EventBus.getDefault().postSticky(new LongTermTaskStateChangedEvent(false, taskCounter));
		throw new RuntimeException(e);
	}

	protected abstract String getTag();

}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/longtermtask/LongTermTask.kt
package org.neidhardt.dynamicsoundboard.longtermtask

import android.os.Handler
import org.greenrobot.eventbus.EventBus
import org.neidhardt.dynamicsoundboard.SoundboardApplication
import org.neidhardt.dynamicsoundboard.misc.Logger
import roboguice.util.SafeAsyncTask

/**
 * File created by eric.neidhardt on 24.03.2015.
 */

abstract class LongTermTask<T> : SafeAsyncTask<T>()
{
	protected abstract val TAG: String

	private val eventBus = EventBus.getDefault()

	@Throws(Exception::class)
	override fun onPreExecute() {
		super.onPreExecute()
		SoundboardApplication.taskCounter += 1
	}

	@Throws(Exception::class)
	override fun onSuccess(result: T) {
		super.onSuccess(result)
		SoundboardApplication.taskCounter -= 1
	}

	@Throws(RuntimeException::class)
	override fun onException(e: Exception) {
		super.onException(e)
		Logger.e(TAG, e.message)
		SoundboardApplication.taskCounter -= 1
		throw RuntimeException(e)
	}
}
